step_by_step_contribute.md

```diff
@@ -0,0 +1,261 @@
+
+    ```sh
+    cd $working_dir # Comes to the directory that you want put the fork in, for example, "cd ~/code"
+    git clone git@github.com:$user/dm.git # Replace "$user" with your GitHub ID
+
+    cd $working_dir/dm
+    git remote add upstream git@github.com:pingcap/dm.git # Adds the upstream repo
+    git remote -v # Confirms that your remote makes sense
+    ```
+
+### Step 3: Setting up your development environment
+
+1. Setting up a Go development environment: [How to Write Go Code](http://golang.org/doc/code.html).
+
+    > **Note:**
+    >
+    > DM uses [`Go Modules`](https://github.com/golang/go/wiki/Modules) to manage dependencies. The version of Go should be **1.13** or above.
+
+    You'll need `GOPATH` defined, and `PATH` modified to access your Go binaries. A
+    common setup is the following but you could always google a setup for your own
+    flavor.
+
+    ```sh
+    export GOPATH=$HOME/go
+    export PATH=$PATH:$GOPATH/bin
+    ```
+
+    Then you can use `make` command to build DM.
+
+    ```sh
+    make build
+    ```
+
+2. Setting up test environment. See [Test Preparations](tests/README.md#Preparations) for more details.
+
+### Step 4: Pick up a challenge program
+
+You can start by finding an existing issue with the
+[UCP](https://github.com/pingcap/dm/labels/challenge-program-2)
+label in the DM repository. These issues are well suited for new contributors.
+
+> **Note:**
+>
+> This section takes challenge program [UCP: retrieve the configuration of the running task from the DM cluster](https://github.com/pingcap/dm/issues/182) as an example. Steps of pick up other challege programs are similar.
+
+Add following comment for the github issue to pick up the challenge program.
+
+    ```
+    /pick-up-challenge
+    ```
+
+### Step 5: Create a new branch
+
+1. Get your local master up-to-date with upstream/master.
+
+    ```
+    cd $working_dir/dm
+    git fetch upstream
+    git checkout master
+    git rebase upstream/master
+    ```
+
+2. Create a new branch based on the master branch.
+
+    ```
+    git checkout -b new-branch-name
+    ```
+
+### Step 6: Edit your code
+
+Edit some code on the `new-branch-name` branch and save your changes to fix the issure. Below is a example to add `get-task-config` command for DM.
+
+1. Update the [proto code](dm/proto/dmmaster.proto) for grpc
+
+    ```
+    // GetTaskCfg implements a rpc method to get task config
+    rpc GetTaskCfg(GetTaskCfgRequest) returns(GetTaskCfgResponse) {
+        ...
+    }
+    
+    // GetTaskCfgRequest is a rpc request for GetTaskCfg
+    message GetTaskCfgRequest {
+        ...
+    }
+    
+    // GetTaskCfgRequest is a rpc response for GetTaskCfg
+    message GetTaskCfgResponse {
+        ...
+    } 
+    ```
+
+2. Generate proto code
+
+    ```sh
+    make generate_proto
+    ```
+
+3. Generate mock code
+
+    ```sh
+    make generate_mock
+    ```
+
+4. Add some error instance for your new command in [error_list](pkg/terror/error_list.go)
```

Comment:
only when we almost finish the development, can we know what error should be added